identity_jose/src/jws/decoder.rs

```diff
@@ -0,0 +1,251 @@
+/// The signature as a slice.
+pub type DecoderSignature<'a> = &'a [u8];
+
+const COMPACT_SEGMENTS: usize = 3;
+
+/// A partially decoded token from a JWS.
+///
+/// Contains the decoded headers and the raw claims.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Token<'a> {
+  pub protected: Option<JwsHeader>,
+  pub unprotected: Option<JwsHeader>,
+  pub claims: Cow<'a, [u8]>,
+}
+
+#[derive(serde::Deserialize)]
+#[serde(deny_unknown_fields)]
+struct JwsSignature<'a> {
+  header: Option<JwsHeader>,
+  protected: Option<&'a str>,
+  signature: &'a str,
+}
+
+#[derive(serde::Deserialize)]
+#[serde(deny_unknown_fields)]
+struct General<'a> {
+  payload: Option<&'a str>,
+  signatures: Vec<JwsSignature<'a>>,
+}
+
+#[derive(serde::Deserialize)]
+#[serde(deny_unknown_fields)]
+struct Flatten<'a> {
+  payload: Option<&'a str>,
+  #[serde(flatten)]
+  signature: JwsSignature<'a>,
+}
+
+// =============================================================================
+// =============================================================================
+
+/// The [`Decoder`] allows decoding a raw JWS into a [`Token`], verifying
+/// the structure of the JWS and its signature.
+///
+/// When attempting to decode a raw JWS, the decoder will check for the expected format, algorithms
+/// and crits which can be set via their respective setters.
+///
+/// This API does not have any cryptography built-in. Rather, signatures are verified
+/// through a closure that is passed to the [`decode`](Decoder::decode) method, so that users can implement
+/// verification for their particular algorithm.
+pub struct Decoder<'b> {
+  /// The expected format of the encoded token.
+  format: JwsFormat,
+  /// A list of permitted signature algorithms.
+  algs: Option<Vec<JwsAlgorithm>>,
+  /// A list of permitted extension parameters.
+  crits: Option<Vec<String>>,
+  /// The detached payload, if using detached content
+  ///
+  /// [More Info](https://tools.ietf.org/html/rfc7515#appendix-F)
+  payload: Option<&'b [u8]>,
+}
+
+impl<'a, 'b> Decoder<'b> {
+  pub fn new() -> Self {
+    Self {
+      format: JwsFormat::Compact,
+
+      algs: None,
+      crits: None,
+      payload: None,
+    }
+  }
+
+  pub fn format(mut self, value: JwsFormat) -> Self {
+    self.format = value;
+    self
+  }
+
+  pub fn algorithm(mut self, value: JwsAlgorithm) -> Self {
+    self.algs.get_or_insert_with(Vec::new).push(value);
+    self
+  }
+
+  pub fn critical(mut self, value: impl Into<String>) -> Self {
+    self.crits.get_or_insert_with(Vec::new).push(value.into());
+    self
+  }
+
+  pub fn payload(mut self, value: &'b [u8]) -> Self {
+    self.payload = Some(value);
+    self
+  }
+
+  /// Decode the given `data` which is a base64url-encoded JWS.
+  ///
+  /// The `verify_fn` closure must be provided to verify signatures on the JWS.
+  /// Only one signature must match in order for the JWS to be considered valid,
+  /// hence `verify_fn` can error on signatures it cannot verify.
```

Comment:
Consider adding a sentence explaining when multiple signatures can and cannot be expected. Also it would be nice to describe in some detail when `verify_fn` must return an error. Something like "verify_fn must return an error if the `alg` parameter in the header describes a cryptographic algorithm `verify_fn` is not designed to handle,  or if signature verification fails".